src/core/positioning/egenioussreceiver.cpp

```diff
@@ -0,0 +1,128 @@
+#include "egenioussreceiver.h"
+
+#include <QHostAddress>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonValue>
+
+EgenioussReceiver::EgenioussReceiver( const QString &address, const int port, QObject *parent )
+  : AbstractGnssReceiver( parent ), mAddress( address ), mPort( port ), mTcpSocket( new QTcpSocket( this ) )
+{
+  setValid( !mAddress.isEmpty() && mPort > 0 );
+}
+
+void EgenioussReceiver::handleConnectDevice()
+{
+  connect( mTcpSocket, &QTcpSocket::readyRead, this, &EgenioussReceiver::onReadyRead );
+  connect( mTcpSocket, &QTcpSocket::errorOccurred, this, &EgenioussReceiver::onErrorOccurred );
+
+  if ( mAddress.isEmpty() || mPort == 0 )
+  {
+    return;
+  }
+
+  mTcpSocket->connectToHost( QHostAddress( mAddress ), mPort, QTcpSocket::ReadWrite );
+
+  if ( !mTcpSocket->waitForConnected( 3000 ) )
+  {
+    mLastError = mTcpSocket->errorString();
+    emit lastErrorChanged( mLastError );
+  }
+  else
+  {
+    mSocketState = QAbstractSocket::ConnectedState;
+    emit socketStateChanged( mSocketState );
+  }
+}
+
+void EgenioussReceiver::handleDisconnectDevice()
+{
+  disconnect( mTcpSocket, &QTcpSocket::readyRead, this, &EgenioussReceiver::onReadyRead );
+  disconnect( mTcpSocket, &QTcpSocket::errorOccurred, this, &EgenioussReceiver::onErrorOccurred );
+
+  if ( mTcpSocket->state() == QAbstractSocket::ConnectedState )
+  {
+    mTcpSocket->disconnectFromHost();
+    mTcpSocket->waitForDisconnected();
+  }
+}
+
+QList<QPair<QString, QVariant>> EgenioussReceiver::details()
+{
+  QList<QPair<QString, QVariant>> dataList;
+
+  QJsonDocument jsonDoc = QJsonDocument::fromJson( payload );
+  QJsonObject jsonObject = jsonDoc.object();
```

Comment:
Don't store the payload as a byte array, store is as a QJsonObject. You already create that object in onReadyRead, let's avoid the double-cost here.

Reply:
Nice :+1: